api/src/main/java/org/pmiops/workbench/api/AuditController.java

```diff
@@ -0,0 +1,138 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import javax.inject.Provider;
+import org.pmiops.workbench.annotations.AuthorityRequired;
+import org.pmiops.workbench.config.WorkbenchConfig;
+import org.pmiops.workbench.db.dao.UserDao;
+import org.pmiops.workbench.model.AuditBigQueryResponse;
+import org.pmiops.workbench.model.Authority;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.RestController;
+
+
+/**
+ * The audit controller is meant for performing offline audit checks. Currently, audit violations
+ * result in SEVERE log statements being written, which may then be alerted on in cloud console.
+ * The audit policy implemented by this API reflects the design outlined here:
+ *   https://docs.google.com/document/d/14HT1GWXHPMaCc9rhCM0y5CglAIY-GgRBwebdZpqwEDs
+ */
+@RestController
+public class AuditController implements AuditApiDelegate {
+
+  private static final Logger log = Logger.getLogger(AuditController.class.getName());
+  private static final String AUDIT_SINK_NAME = "cdr_audit_logs";
+  // How many days into the past (including today) logs should be checked. This could become a
+  // request parameter if the need arises.
+  private static final int AUDIT_DAY_RANGE = 7;
+  // BigQuery log sink table names are have a suffix like "20170103", per
+  // https://cloud.google.com/logging/docs/export/using_exported_logs#table_organization
+  private static final DateTimeFormatter auditTableNameDateFormatter =
+      new DateTimeFormatterBuilder()
+        .appendValue(ChronoField.YEAR, 4)
+        .appendValue(ChronoField.MONTH_OF_YEAR, 2)
+        .appendValue(ChronoField.DAY_OF_MONTH, 2)
+        .toFormatter();
+
+  private final Clock clock;
+  private final BigQueryService bigQueryService;
+  private final UserDao userDao;
+  private final Provider<WorkbenchConfig> workbenchConfigProvider;
+
+  @Autowired
+  AuditController(
+      Clock clock,
+      BigQueryService bigQueryService,
+      UserDao userDao,
+      Provider<WorkbenchConfig> workbenchConfigProvider) {
+    this.clock = clock;
+    this.bigQueryService = bigQueryService;
+    this.userDao = userDao;
+    this.workbenchConfigProvider = workbenchConfigProvider;
+  }
+
+  @VisibleForTesting
+  static String auditTableSuffix(Instant now, int daysAgo) {
+    Instant target = now.minus(daysAgo, ChronoUnit.DAYS);
+    return auditTableNameDateFormatter.withZone(ZoneId.of("UTC")).format(target);
+  }
+
+  private static String auditSql(String cdrProjectId, List<String> tableSuffixes) {
+    // "jobInsertResponse" appears to always be included, despite whether or not job request
+    // metadata was included (i.e. for jobs running in other projects).
+    String tableWildcard = String.format(
+        "%s.%s.cloudaudit_googleapis_com_data_access_*", cdrProjectId, AUDIT_SINK_NAME);
+    return String.format(
+        "SELECT\n" +
```

Comment:
Can we get IP address out of this, too?

At some point we'd discussed alerting if the CDR was accessed from an IP address that didn't appear to be in a notebook or workbench server. I can't remember where we left that...